src/goto-programs/goto_program.h

```diff
@@ -199,13 +199,21 @@ class goto_programt
     const code_declt &get_decl() const
     {
       PRECONDITION(is_decl());
-      return to_code_decl(code);
+      const auto &decl = expr_checked_cast<code_declt>(code);
+      INVARIANT(
+        !decl.initial_value(),
+        "code_declt in goto program may not contain initialization.");
+      return decl;
```

Comment:
While the `INVARIANT` in `set_decl` is probably the right thing, this one may better be placed in `goto_programt::instructiont::validate`.

Reply:
Given that this PR is now approved, I am going to merge as is and clean up this as a follow-up PR.